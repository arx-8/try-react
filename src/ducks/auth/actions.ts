import { APP_NAME } from "src/constants/App"
import actionCreatorFactory from "typescript-fsa"

export const ActionTypes = {
  LOGIN: "auth/LOGIN",
  LOGOUT: "auth/LOGOUT",
}

const create = actionCreatorFactory(APP_NAME)

export const login = create(ActionTypes.LOGIN)

export const logout = create(ActionTypes.LOGOUT)
