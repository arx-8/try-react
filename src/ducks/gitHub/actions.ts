import { ActionHandler, ActionWithPayloadHandler } from "src/types/ReduxTypes"
import {
  FetchMemberError,
  FetchMemberResponse,
  FetchMembersParams,
} from "./types"

export const ActionTypes = {
  FETCH_MEMBERS_START: "app/gitHub/FETCH_MEMBERS_START",
  FETCH_MEMBERS_SUCCEED: "app/gitHub/FETCH_MEMBERS_SUCCEED",
  FETCH_MEMBERS_FAIL: "app/gitHub/FETCH_MEMBERS_FAIL",
  INITIALIZE_MEMBERS: "app/gitHub/INITIALIZE_MEMBERS",
} as const

export const fetchMembersStart: ActionWithPayloadHandler<
  typeof ActionTypes.FETCH_MEMBERS_START,
  FetchMembersParams
> = (payload) => ({
  type: ActionTypes.FETCH_MEMBERS_START,
  payload,
})

export const fetchMembersSucceed: ActionWithPayloadHandler<
  typeof ActionTypes.FETCH_MEMBERS_SUCCEED,
  FetchMembersParams & FetchMemberResponse
> = (payload) => ({
  type: ActionTypes.FETCH_MEMBERS_SUCCEED,
  payload,
})

export const fetchMembersFail: ActionWithPayloadHandler<
  typeof ActionTypes.FETCH_MEMBERS_FAIL,
  FetchMembersParams & FetchMemberError
> = (payload) => ({
  type: ActionTypes.FETCH_MEMBERS_FAIL,
  payload,
})

export const initializeMembers: ActionHandler<
  typeof ActionTypes.INITIALIZE_MEMBERS
> = () => ({
  type: ActionTypes.INITIALIZE_MEMBERS,
})

export type Action =
  | ReturnType<typeof fetchMembersStart>
  | ReturnType<typeof fetchMembersSucceed>
  | ReturnType<typeof fetchMembersFail>
  | ReturnType<typeof initializeMembers>
