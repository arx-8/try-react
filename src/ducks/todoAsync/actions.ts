import {
  callDeleteTodo,
  CallDeleteTodoReq,
  callGetAllTodos,
  callGetTodo,
  CallGetTodoReq,
  callPostTodo,
  CallPostTodoReq,
  callPutTodo,
  CallPutTodoReq,
} from "data/repository/TodoRepository"
import {
  SerializableError,
  toSerializableError,
} from "domain/errors/SerializableError"
import { Todo, TodoId, VisibilityFilter } from "domain/models/Todo"
import debounce from "lodash/debounce"
import { ThunkActionCreatorReturn } from "types/ReduxTypes"
import actionCreatorFactory from "typescript-fsa"
import { TodoAsyncDispatch } from "./types"

const TargetDomain = "TODO_ASYNC"

export enum ActionTypes {
  ADD_TODO = "ADD_TODO",
  CHANGE_TODO_STATUS = "CHANGE_TODO_STATUS",
  DELETE_TODO = "DELETE_TODO",
  FETCH_ALL_TODOS = "FETCH_ALL_TODOS",
  FETCH_TODO = "FETCH_TODO",
  SET_VISIBILITY_FILTER = "SET_VISIBILITY_FILTER",
  SET_EDIT_TARGET_ID = "SET_EDIT_TARGET_ID",
}

const create = actionCreatorFactory(TargetDomain)

const addTodo = create.async<CallPostTodoReq, TodoId, SerializableError>(
  ActionTypes.ADD_TODO
)

const addTodoRequest = (
  params: CallPostTodoReq
): ThunkActionCreatorReturn<TodoAsyncDispatch> => {
  return async (dispatch) => {
    dispatch(addTodo.started(params))
    try {
      const id = await callPostTodo(params)
      dispatch(addTodo.done({ params, result: id }))
      dispatch(fetchAllTodosRequestDebounce())
    } catch (error) {
      dispatch(addTodo.failed({ error: toSerializableError(error), params }))
    }
  }
}

const updateTodo = create.async<CallPutTodoReq, TodoId, SerializableError>(
  ActionTypes.CHANGE_TODO_STATUS
)

const updateTodoRequest = (
  params: CallPutTodoReq
): ThunkActionCreatorReturn<TodoAsyncDispatch> => {
  return async (dispatch) => {
    dispatch(updateTodo.started(params))
    try {
      const id = await callPutTodo(params)
      dispatch(updateTodo.done({ params, result: id }))
      dispatch(fetchAllTodosRequestDebounce())
    } catch (error) {
      dispatch(updateTodo.failed({ error: toSerializableError(error), params }))
    }
  }
}

const deleteTodo = create.async<CallDeleteTodoReq, void, SerializableError>(
  ActionTypes.DELETE_TODO
)

const deleteTodoRequest = (
  params: CallDeleteTodoReq
): ThunkActionCreatorReturn<TodoAsyncDispatch> => {
  return async (dispatch) => {
    dispatch(deleteTodo.started(params))
    try {
      await callDeleteTodo(params)
      dispatch(deleteTodo.done({ params }))
      dispatch(fetchAllTodosRequestDebounce())
    } catch (error) {
      dispatch(deleteTodo.failed({ error: toSerializableError(error), params }))
    }
  }
}

const fetchTodo = create.async<CallGetTodoReq, Todo, SerializableError>(
  ActionTypes.FETCH_TODO
)

const fetchTodoRequest = (
  params: CallGetTodoReq
): ThunkActionCreatorReturn<TodoAsyncDispatch> => {
  return async (dispatch) => {
    dispatch(fetchTodo.started(params))
    try {
      dispatch(fetchTodo.done({ result: await callGetTodo(params), params }))
    } catch (error) {
      dispatch(
        fetchTodo.failed({
          error: toSerializableError(error, error.code),
          params,
        })
      )
    }
  }
}

const fetchAllTodos = create.async<void, Todo[], SerializableError>(
  ActionTypes.FETCH_ALL_TODOS
)

/**
 * debounceなしで実行
 */
const fetchAllTodosRequest = (): ThunkActionCreatorReturn<
  TodoAsyncDispatch
> => {
  return fetchAllTodosRequestActInner
}

/**
 * debounceありで実行
 */
const fetchAllTodosRequestDebounce = (): ThunkActionCreatorReturn<
  TodoAsyncDispatch
> => {
  return fetchAllTodosRequestDebounceMemo
}

const fetchAllTodosRequestActInner = async (
  dispatch: TodoAsyncDispatch
): Promise<any> => {
  dispatch(fetchAllTodos.started())
  try {
    dispatch(fetchAllTodos.done({ result: await callGetAllTodos() }))
  } catch (error) {
    dispatch(fetchAllTodos.failed({ error: toSerializableError(error) }))
  }
}

/**
 * debounce可能にするためのメモ化
 */
const fetchAllTodosRequestDebounceMemo = debounce(
  fetchAllTodosRequestActInner,
  500
)

const setEditTargetId = create<{
  editTargetId?: TodoId
}>(ActionTypes.SET_EDIT_TARGET_ID)

const setVisibilityFilter = create<{
  visibilityFilter: VisibilityFilter
}>(ActionTypes.SET_VISIBILITY_FILTER)

/**
 * plain actions
 */
export const actions = {
  addTodo,
  deleteTodo,
  fetchAllTodos,
  fetchTodo,
  setEditTargetId,
  setVisibilityFilter,
  updateTodo,
}

/**
 * thunk actions
 */
export const requestActions = {
  addTodoRequest,
  deleteTodoRequest,
  fetchAllTodosRequest,
  fetchAllTodosRequestDebounce,
  fetchTodoRequest,
  updateTodoRequest,
}
