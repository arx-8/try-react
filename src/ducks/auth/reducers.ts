import { AuthToken } from "domain/models/Auth"
import produce from "immer"
import { reducerWithInitialState } from "typescript-fsa-reducers"
import { actions } from "./actions"

export type State = Readonly<{
  authToken?: AuthToken
}>

export const initialState: State = {
  authToken: undefined,
}

export const reducer = reducerWithInitialState(initialState)
  .case(actions.login, (state) => {
    return produce(state, (draft) => {
      draft.authToken = "any auth token" as AuthToken
    })
  })
  .case(actions.logout, (state) => {
    return produce(state, (draft) => {
      draft.authToken = undefined
    })
  })
  .build()
