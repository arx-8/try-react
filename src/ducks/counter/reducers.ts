import * as actions from "./actions"
import { CounterActionType } from "./types"
import { handleActions, ReduxCompatibleReducer } from "redux-actions"
import produce from "immer"

export type CounterState = Readonly<{
  count: number
  defaultAmount: number
}>

// redux-actions の型がうまく書けない。redux-actions はだいぶダメな感じ。
type Action =
  | ReturnType<
      | typeof actions.add
      | typeof actions.changeDefaultAmount
      | typeof actions.decrement
      | typeof actions.increment
    >
  | any

export const initialState: CounterState = {
  count: 0,
  defaultAmount: 1,
}

export const reducer: ReduxCompatibleReducer<
  CounterState,
  Action
> = handleActions(
  {
    [CounterActionType.ADD]: (state, action) => {
      return produce(state, (draft) => {
        draft.count = state.count + action.payload.count
      })
    },
    [CounterActionType.DECREMENT]: (state) => {
      return produce(state, (draft) => {
        draft.count = state.count - 1
      })
    },
    [CounterActionType.INCREMENT]: (state) => {
      return produce(state, (draft) => {
        draft.count = state.count + 1
      })
    },
    [CounterActionType.CHANGE_DEFAULT_AMOUNT]: (state, action) => {
      return produce(state, (draft) => {
        draft.defaultAmount = action.payload.defaultAmount
      })
    },
  },
  initialState
)
