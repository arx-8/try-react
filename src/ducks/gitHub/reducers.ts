import { Action, ActionTypes } from "./actions"
import { GitHubState } from "./types"
import { Reducer } from "redux"
import produce from "immer"

export const initialState: GitHubState = {
  users: [],
  isLoading: false,
  error: undefined,
}

export const reducer: Reducer<GitHubState, Action> = (
  state = initialState,
  action
) => {
  switch (action.type) {
    case ActionTypes.FETCH_MEMBERS_START:
      return produce(state, (draft) => {
        draft.isLoading = true
      })

    case ActionTypes.FETCH_MEMBERS_SUCCEED:
      return produce(state, (draft) => {
        const { users } = action.payload
        draft.users = users
        draft.isLoading = false
      })

    case ActionTypes.FETCH_MEMBERS_FAIL:
      return produce(state, (draft) => {
        const { error } = action.payload
        draft.error = error
        draft.isLoading = false
      })

    case ActionTypes.INITIALIZE_MEMBERS:
      return produce(state, (draft) => {
        draft.users = []
      })

    default: {
      // case の定義忘れ防止のため
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      const _: never = action
      return state
    }
  }
}
