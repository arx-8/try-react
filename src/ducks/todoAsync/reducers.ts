import { Todo, VisibilityFilter, TodoId } from "domain/models/Todo"
import produce from "immer"
import { reducerWithInitialState } from "typescript-fsa-reducers"
import { actions } from "./actions"

export type State = Readonly<{
  todoList: Todo[]
  visibilityFilter: VisibilityFilter
  loading: {
    all: boolean
    add: boolean
    ids: TodoId[]
  }
  errorMessage?: string
}>

export const initialState: State = {
  todoList: [],
  visibilityFilter: "all",
  loading: {
    all: false,
    add: false,
    ids: [],
  },
  errorMessage: undefined,
}

export const reducer = reducerWithInitialState(initialState)
  /**
   * shared
   */
  .cases(
    [actions.changeTodoStatus.done, actions.deleteTodo.done],
    (state, payload) => {
      return produce(state, (draft) => {
        draft.loading.ids = draft.loading.ids.filter(
          (id) => id !== payload.params.id
        )
        draft.errorMessage = undefined
      })
    }
  )
  .cases(
    [actions.changeTodoStatus.failed, actions.deleteTodo.failed],
    (state, payload) => {
      return produce(state, (draft) => {
        draft.loading.ids = draft.loading.ids.filter(
          (id) => id !== payload.params.id
        )
        draft.errorMessage = payload.error.message
      })
    }
  )
  /**
   * changeTodoStatus
   */
  .case(actions.changeTodoStatus.started, (state, payload) => {
    return produce(state, (draft) => {
      const { id, label, status } = payload
      draft.loading.ids.push(id)

      // Optimistic Updates
      const t = draft.todoList.find((t) => t.id === id)!
      if (label) {
        t.label = label
      }
      if (status) {
        t.status = status
      }
    })
  })
  /**
   * deleteTodo
   */
  .case(actions.deleteTodo.started, (state, payload) => {
    return produce(state, (draft) => {
      const { id } = payload
      draft.loading.ids.push(id)

      // Optimistic Updates
      draft.todoList = draft.todoList.filter((t) => t.id !== id)!
    })
  })
  /**
   * addTodo
   */
  .case(actions.addTodo.started, (state) => {
    return produce(state, (draft) => {
      draft.loading.add = true
    })
  })
  .case(actions.addTodo.done, (state) => {
    return produce(state, (draft) => {
      draft.loading.add = false
      draft.errorMessage = undefined
    })
  })
  .case(actions.addTodo.failed, (state, payload) => {
    return produce(state, (draft) => {
      draft.loading.add = false
      draft.errorMessage = payload.error.message
    })
  })
  /**
   * fetchAllTodos
   */
  .case(actions.fetchAllTodos.started, (state) => {
    return produce(state, (draft) => {
      draft.loading.all = true
    })
  })
  .case(actions.fetchAllTodos.done, (state, payload) => {
    return produce(state, (draft) => {
      const { result } = payload
      draft.todoList = result
      draft.loading.all = false
      draft.errorMessage = undefined
    })
  })
  .case(actions.fetchAllTodos.failed, (state, payload) => {
    return produce(state, (draft) => {
      draft.loading.all = false
      draft.errorMessage = payload.error.message
    })
  })
  /**
   * setVisibilityFilter
   */
  .case(actions.setVisibilityFilter, (state, payload) => {
    return produce(state, (draft) => {
      const { visibilityFilter } = payload
      draft.visibilityFilter = visibilityFilter
    })
  })
  .build()
