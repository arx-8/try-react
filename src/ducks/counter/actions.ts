import { createAction } from "redux-actions"
import { CounterActionType } from "./types"
import { CounterState } from "./reducers"

export const add = createAction(
  CounterActionType.ADD,
  (amount: number): Partial<CounterState> => ({
    count: amount,
  })
)

export const changeDefaultAmount = createAction(
  CounterActionType.CHANGE_DEFAULT_AMOUNT,
  (defaultAmount: number): Partial<CounterState> => ({ defaultAmount })
)

export const decrement = createAction(CounterActionType.DECREMENT)

export const increment = createAction(CounterActionType.INCREMENT)
