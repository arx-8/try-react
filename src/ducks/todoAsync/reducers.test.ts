import { initialState, reducer } from "./reducers"
import { actions } from "./actions"
import { TodoId } from "domain/models/Todo"
import produce from "immer"

describe("updateTodo", () => {
  it("updateTodo", async () => {
    // ## Arrange ##
    const prevState = produce(initialState, (draft) => {
      draft.todoList = [
        {
          id: "id0" as TodoId,
          label: "old label 0",
          status: "active",
        },
        {
          id: "id1" as TodoId,
          label: "old label 1",
          status: "active",
        },
      ]
    })

    // ## Act ##
    const result = reducer(
      prevState,
      actions.updateTodo.started({
        id: "id1" as TodoId,
        label: "NEW LABEL 1",
        status: "completed",
      })
    )

    // ## Assert ##
    expect(result).toMatchSnapshot()
  })
})
