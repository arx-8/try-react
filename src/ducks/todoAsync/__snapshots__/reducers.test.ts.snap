// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updateTodo updateTodo 1`] = `
Object {
  "editTargetId": undefined,
  "errorMessage": undefined,
  "loading": Object {
    "add": false,
    "all": false,
    "ids": Array [
      "id1",
    ],
  },
  "todoList": Array [
    Object {
      "id": "id0",
      "label": "old label 0",
      "status": "active",
    },
    Object {
      "id": "id1",
      "label": "NEW LABEL 1",
      "status": "completed",
    },
  ],
  "visibilityFilter": "all",
}
`;
